Nil pointer dereference due to malformed CronWorkflow. Workflow-controller crashlooping

### Pre-requisites

- [x] I have double-checked my configuration
- [x] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [x] I have searched existing issues and could not find a match for this bug
- [x] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

Controller version: `argoproj/workflow-controller:v3.7.2`

**Current Behavior:**

The workflow-controller panics and enters a crash loop when it attempts to process a CronWorkflow that is malformed (likely with a missing or nil field within its workflowSpec).

Instead of marking the CronWorkflow as invalid and logging an error, the controller panics with runtime error: "invalid memory address or nil pointer dereference." and enters a crashloop.

**Expected Behavior:**
The controller should gracefully handle malformed CronWorkflow resources. It should:

- Fail to validate the resource.
- Update the .status of the CronWorkflow with a condition indicating the validation error.
- Log a clear error message (like "Failed to validate CronWorkflow...").

Continue running without crashing to manage other valid resources.

Logs from the workflow controller:


Status object from the problematic CronWorkflow resource:
```
Status:                                                                                                                                                                                           
  Conditions:                                                                                                                                                                                     
    Message:  cannot validate Workflow: template name 'redacted' undefined                                                                                                                
    Status:   True                                                                                                                                                                                
    Type:     SpecError                                                                                                                                                                           
  Failed:     0                                                                                                                                                                                   
  Phase:                                                                                                                                                                                          
  Succeeded:  0      
```

### Version(s)

v3.7.2

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflow that uses private images.

```YAML
Create a Cronworkflow with a missing template?
```

### Logs from the workflow controller

```text
time="2025-10-22T21:42:09.319Z" level=info msg="Processing redacted/redacted" cronWorkflow=redacted/redacted
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x90 pc=0x1ee75c1]

goroutine 236 [running]:
k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x343f9f0, 0xc0000120e0}, {0x2aa4980, 0x4e1bde0}, {0x0, 0x0, 0x440d80?})
    /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:114 +0x1a9
k8s.io/apimachinery/pkg/util/runtime.HandleCrashWithContext({0x343f9f0, 0xc0000120e0}, {0x0, 0x0, 0x0})
    /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:78 +0x5a
panic({0x2aa4980?, 0x4e1bde0?})
    /usr/local/go/src/runtime/panic.go:792 +0x132
k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x343f138, 0x4eab0e0}, {0x2aa4980, 0x4e1bde0}, {0x4e19990, 0x1, 0xc0005bf300?})
    /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:114 +0x1a9
k8s.io/apimachinery/pkg/util/runtime.HandleCrashWithContext({0x343f138, 0x4eab0e0}, {0x4e19990, 0x1, 0x1})
    /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:78 +0x5a
panic({0x2aa4980?, 0x4e1bde0?})
    /usr/local/go/src/runtime/panic.go:792 +0x132
[github.com/argoproj/argo-workflows/v3/workflow/templateresolution.(*Context).addPodMetadata](https://github.com/argoproj/argo-workflows/v3/workflow/templateresolution.(*Context).addPodMetadata)(...)
    /go/src/[github.com/argoproj/argo-workflows/workflow/templateresolution/context.go:295](https://github.com/argoproj/argo-workflows/workflow/templateresolution/context.go:295)
[github.com/argoproj/argo-workflows/v3/workflow/templateresolution.(*Context).GetTemplateByName(0xc001e5d0e8](https://github.com/argoproj/argo-workflows/v3/workflow/templateresolution.(*Context).GetTemplateByName(0xc001e5d0e8), {0xc0014d4c90, 0x10})
    /go/src/[github.com/argoproj/argo-workflows/workflow/templateresolution/context.go:122](https://github.com/argoproj/argo-workflows/workflow/templateresolution/context.go:122) +0xe1
[github.com/argoproj/argo-workflows/v3/workflow/validate.(*templateValidationCtx).validateTemplateHolder(0xc001ef2220](https://github.com/argoproj/argo-workflows/v3/workflow/validate.(*templateValidationCtx).validateTemplateHolder(0xc001ef2220), {0x3446620, 0xc0020d63c0}, 0xc001e5d0e8, {0x3413088, 0xc002ca0d50}, 0x0)
    /go/src/[github.com/argoproj/argo-workflows/workflow/validate/validate.go:565](https://github.com/argoproj/argo-workflows/workflow/validate/validate.go:565) +0xab
[github.com/argoproj/argo-workflows/v3/workflow/validate.ValidateWorkflow](https://github.com/argoproj/argo-workflows/v3/workflow/validate.ValidateWorkflow)({0x34037e0, 0xc001ef2180}, {0x3403800, 0xc002e037b0}, 0xc002f02488, 0x0, {0x0, 0x0, 0x0, 0x0})
    /go/src/[github.com/argoproj/argo-workflows/workflow/validate/validate.go:257](https://github.com/argoproj/argo-workflows/workflow/validate/validate.go:257) +0x219e
[github.com/argoproj/argo-workflows/v3/workflow/validate.ValidateCronWorkflow](https://github.com/argoproj/argo-workflows/v3/workflow/validate.ValidateCronWorkflow)({0x343f058?, 0xc002ca0ba0?}, {0x34037e0, 0xc001ef2180}, {0x3403800, 0xc002e037b0}, 0xc00227c408, 0x0)
    /go/src/[github.com/argoproj/argo-workflows/workflow/validate/validate.go:404](https://github.com/argoproj/argo-workflows/workflow/validate/validate.go:404) +0x34d
[github.com/argoproj/argo-workflows/v3/workflow/cron.(*cronWfOperationCtx).validateCronWorkflow(0xc000630630](https://github.com/argoproj/argo-workflows/v3/workflow/cron.(*cronWfOperationCtx).validateCronWorkflow(0xc000630630), {0x343f058, 0xc002ca0ba0})
    /go/src/[github.com/argoproj/argo-workflows/workflow/cron/operator.go:146](https://github.com/argoproj/argo-workflows/workflow/cron/operator.go:146) +0x15f
[github.com/argoproj/argo-workflows/v3/workflow/cron.(*Controller).processNextCronItem(0xc000246460](https://github.com/argoproj/argo-workflows/v3/workflow/cron.(*Controller).processNextCronItem(0xc000246460), {0x343f138, 0x4eab0e0})
    /go/src/[github.com/argoproj/argo-workflows/workflow/cron/controller.go:182](https://github.com/argoproj/argo-workflows/workflow/cron/controller.go:182) +0x8c5
[github.com/argoproj/argo-workflows/v3/workflow/cron.(*Controller).runCronWorker](https://github.com/argoproj/argo-workflows/v3/workflow/cron.(*Controller).runCronWorker)(...)
    /go/src/[github.com/argoproj/argo-workflows/workflow/cron/controller.go:136](https://github.com/argoproj/argo-workflows/workflow/cron/controller.go:136)
```

### Logs from in your workflow's wait container

```text
does not have a wait container?
```

--- a/workflow/templateresolution/context.go
+++ b/workflow/templateresolution/context.go
@@ -148,12 +148,13 @@ func (tplCtx *TemplateContext) GetTemplateFromRef(ctx context.Context, tmplRef *
 
 	template = wftmpl.GetTemplateByName(tmplRef.Template)
 
-	podMetadata := wftmpl.GetPodMetadata()
-	tplCtx.addPodMetadata(podMetadata, template)
-
 	if template == nil {
 		return nil, errors.Errorf(errors.CodeNotFound, "template %s not found in workflow template %s", tmplRef.Template, tmplRef.Name)
 	}
+
+	podMetadata := wftmpl.GetPodMetadata()
+	tplCtx.addPodMetadata(podMetadata, template)
+
 	return template.DeepCopy(), nil
 }
 

--- a/workflow/templateresolution/context_test.go
+++ b/workflow/templateresolution/context_test.go
@@ -362,3 +362,56 @@ func TestOnWorkflowTemplate(t *testing.T) {
 	tmpl := newCtx.tmplBase.GetTemplateByName("whalesay")
 	assert.NotNil(t, tmpl)
 }
+
+// TestGetTemplateFromRefWithPodMetadataAndMissingTemplate tests the bug where
+// GetTemplateFromRef causes a nil pointer dereference when:
+// 1. A WorkflowTemplate has podMetadata defined
+// 2. A templateRef references a template name that doesn't exist in that WorkflowTemplate
+func TestGetTemplateFromRefWithPodMetadataAndMissingTemplate(t *testing.T) {
+	ctx := logging.TestContext(t.Context())
+	wfClientset := fakewfclientset.NewSimpleClientset()
+
+	// Create a WorkflowTemplate with podMetadata but without the template "nonexistent"
+	workflowTemplateWithPodMetadata := `
+apiVersion: argoproj.io/v1alpha1
+kind: WorkflowTemplate
+metadata:
+  name: template-with-podmetadata
+spec:
+  podMetadata:
+    labels:
+      example-label: example-value
+    annotations:
+      example-annotation: example-value
+  templates:
+  - name: existing-template
+    container:
+      image: alpine:latest
+      command: [echo, hello]
+`
+
+	err := createWorkflowTemplate(ctx, wfClientset, workflowTemplateWithPodMetadata)
+	require.NoError(t, err)
+
+	// Create a base workflow template to use as context
+	baseWftmpl := unmarshalWftmpl(baseWorkflowTemplateYaml)
+	log := logging.RequireLoggerFromContext(ctx)
+	tplCtx := NewContextFromClientSet(
+		wfClientset.ArgoprojV1alpha1().WorkflowTemplates(metav1.NamespaceDefault),
+		wfClientset.ArgoprojV1alpha1().ClusterWorkflowTemplates(),
+		baseWftmpl,
+		nil,
+		log,
+	)
+
+	// Try to get a template that doesn't exist from a WorkflowTemplate that HAS podMetadata
+	tmplRef := wfv1.TemplateRef{
+		Name:     "template-with-podmetadata",
+		Template: "nonexistent-template",
+	}
+
+	_, err = tplCtx.GetTemplateFromRef(ctx, &tmplRef)
+
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "template nonexistent-template not found in workflow template template-with-podmetadata")
+}

